src/main/java/org/folio/services/instance/InstanceS3Service.java

```diff
@@ -0,0 +1,246 @@
+    "buildPrecedingSucceedingTitlesCql:: Failed to build CQL wrapper for preceding and succeeding titles search";
+  private static final String INSTANCES_PARALLEL_UPSERT_COUNT_PARAM = "bulk-processing.parallel.upsert.count";
+  private static final String DEFAULT_INSTANCES_PARALLEL_UPSERT_COUNT = "10";
+  private static final String PRECEDING_SUCCEEDING_TITLES_BY_INSTANCE_IDS_CQL =
+    "succeedingInstanceId==(%1$s) or precedingInstanceId==(%1$s)";
+  private static final String PRECEDING_SUCCEEDING_TITLE_TABLE = "preceding_succeeding_title";
+  private static final String PRECEDING_TITLES_FIELD = "precedingTitles";
+  private static final String SUCCEEDING_TITLES_FIELD = "succeedingTitles";
+  private static final String OR_OPERATOR = " or ";
+  private static final String ID_FIELD = "id";
+
+  private final FolioS3ClientFactory folioS3ClientFactory;
+  private final Vertx vertx;
+  private final FolioS3Client folioS3Client;
+  private final InstanceService instanceService;
+  private final PostgresClient postgresClient;
+  private final InstanceRepository instanceRepository;
+  private final int instancesParallelUpsertLimit;
+
+  public InstanceS3Service(FolioS3ClientFactory folioS3ClientFactory, Vertx vertx, Map<String, String> okapiHeaders) {
+    this.instancesParallelUpsertLimit = Integer.parseInt(
+      System.getProperty(INSTANCES_PARALLEL_UPSERT_COUNT_PARAM, DEFAULT_INSTANCES_PARALLEL_UPSERT_COUNT));
+    this.vertx = vertx;
+    this.folioS3ClientFactory = folioS3ClientFactory;
+    this.folioS3Client = folioS3ClientFactory.getFolioS3Client();
+    this.instanceService = new InstanceService(vertx.getOrCreateContext(), okapiHeaders);
+    this.instanceRepository = new InstanceRepository(vertx.getOrCreateContext(), okapiHeaders);
+    this.postgresClient = PgUtil.postgresClient(vertx.getOrCreateContext(), okapiHeaders);
+  }
+
+  public Future<InstanceBulkResponse> processInstances(InstanceBulkRequest bulkRequest) {
+    log.debug("processInstances:: Processing bulk instances request, filename: '{}'", bulkRequest.getRecordsFileName());
+    return loadInstances(bulkRequest)
+      .compose(instances -> upsert(instances, bulkRequest))
+      .onFailure(e -> log.warn("Failed to process instances bulk request, filename: '{}'",
+        bulkRequest.getRecordsFileName(), e));
+  }
+
+  private Future<List<JsonObject>> loadInstances(InstanceBulkRequest bulkRequest) {
+    return vertx.executeBlocking(() -> {
+      FolioS3Client s3Client = folioS3ClientFactory.getFolioS3Client();
+      InputStream inputStream = s3Client.read(bulkRequest.getRecordsFileName());
+
+      try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+        return reader.lines()
+          .map(JsonObject::new)
+          .toList();
+      }
+    });
+  }
+
+  private Future<InstanceBulkResponse> upsert(List<JsonObject> instances, InstanceBulkRequest bulkRequest) {
```

Comment:
@RuslanLavrov  I think it would be better to make a different method name, to not overload them. For example, here rename to "upsertInstances". And the next method - "createInstanceAndUpdatePrecedingSucceedingTitles". Or smth like this. What do you think?

Reply:
changed